### PR TITLE
Quasi-Newton methods

### PR DESCRIPTION
**Pull request by msricher**

### CODE REVIEW

File: flik/jacobian.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    Analytic Jacobian class. Holds as a parent class
+    for the FiniteDiffJacobian class, which is intended for
+    approximation-based Jacobian methods. Includes
+    static methods, update_goodbroyden and update_badbroyden,
+    intended to be used to update the jacobian.
+
+    A callable object that calls the jacobian of a function,
+    all given by the user.
+    """
+    def __init__(self, jac=None):
```
Comment (msricher) [inline flik/jacobian.py]: `jac` argument shouldn't have the option to be `None`.  It should be a regular argument.

File: flik/nonlinear.py [hunk 2]
```diff
@@ -98,19 +113,38 @@ def newton_solve(f, J, x_0, eps=1.0e-6, maxiter=100):
...
     b = f(x_0)
 
+    # Choose the step function
+    if method is ("analytic" or "good_broyden"):
+        step_function = step_linear
+    else: 
+        step_function = step_inv
+        A = np.linalg.inv(A)
+
+    # Choose the update function
```
Comment (msricher) [inline flik/nonlinear.py]: Use `==` instead of `is` for string comparisons

File: flik/step.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    Raises
+    ------
+    TypeError
+        If an argument of an invalid type or shape is passed.
+    ValueError
+        If an argument passed has an unreasonable value.
+
+    """
+    # Check f and J return types
```
Comment (msricher) [inline flik/step.py]: Don't do input checks inside functions that will be called in a loop.  This function will be called by `newton_solve`, which has already done input checking.

File: flik/step.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
```
Comment (msricher) [inline flik/step.py]: See comment on `step_linear`

File: flik/step.py (hunk 3 above)
Comment (Ali-Tehrani) [reply flik/step.py]: Fix in Xiao's branch addoption.

File: flik/step.py (hunk 4 above)
Comment (Ali-Tehrani) [reply flik/step.py]: Fix in Xiao's branch addoption

File: flik/jacobian.py (hunk 1 above)
Comment (Ali-Tehrani) [reply flik/jacobian.py]: Fixed in jacobian class.

File: flik/nonlinear.py (hunk 2 above)
Comment (xiaominhc) [reply flik/nonlinear.py]: fixed in branch changetests

**Pull request merged**